File: examples/selected_transform_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct square_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename tsl::reg_param<Vec>::type value) const {
        return tsl::mul<Vec>(value, value);
    }
};

struct add_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::add<Vec>(left, right);
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 13) % 47) - 23);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 7) % 31) - 15);
    }
}

std::vector<std::size_t> make_selection(std::size_t count) {
    std::vector<std::size_t> indices;
    for (std::size_t i = count; i > 0; --i) {
        const std::size_t row = i - 1;
        if ((row % 3) == 0 || (row % 5) == 0) {
            indices.push_back(row);
        }
    }
    return indices;
}

template <std::size_t ParallelN, std::size_t Scale = 0>
bool run_selected_transform_pointer_case() {
    constexpr std::size_t count = 1003;
    constexpr std::int32_t sentinel = 7654321;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);
    const auto indices = make_selection(count);

    std::vector<std::int32_t> output(count, sentinel);
    tsl::algo::transform_selected_unary<ParallelN, Scale>(
        square_op{},
        left.data(),
        indices.data(),
        output.data(),
        indices.size());
    for (std::size_t i = 0; i < indices.size(); ++i) {
        const auto value = left[indices[i]];
        if (output[i] != value * value) {
            return false;
        }
    }
    for (std::size_t i = indices.size(); i < count; ++i) {
        if (output[i] != sentinel) {
            return false;
        }
    }

    output.assign(count, sentinel);
    tsl::algo::transform_selected_binary<ParallelN, Scale>(
        add_op{},
        left.data(),
        right.data(),
        indices.data(),
        output.data(),
        indices.size());
    for (std::size_t i = 0; i < indices.size(); ++i) {
        const auto row = indices[i];
        if (output[i] != left[row] + right[row]) {
            return false;
        }
    }
    for (std::size_t i = indices.size(); i < count; ++i) {
        if (output[i] != sentinel) {
            return false;
        }
    }

    return true;
}

template <std::size_t ParallelN>
bool run_selected_transform_range_case() {
    constexpr std::size_t count = 257;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);
    const auto indices = make_selection(count);

    std::vector<std::int32_t> output(indices.size());
    tsl::algo::transform_selected_binary<ParallelN>(
        add_op{},
        left,
        right,
        indices,
        output);
    for (std::size_t i = 0; i < indices.size(); ++i) {
        const auto row = indices[i];
        if (output[i] != left[row] + right[row]) {
            return false;
        }
    }
    return true;
}

template <std::size_t ParallelN>
bool run_selected_transform_cases() {
    return run_selected_transform_pointer_case<ParallelN>() &&
           run_selected_transform_range_case<ParallelN>();
}

int main() {
    if (!run_selected_transform_cases<1>()) {
        return 1;
    }
    if (!run_selected_transform_cases<4>()) {
        return 2;
    }
    if (!run_selected_transform_cases<8>()) {
        return 3;
    }
    if (!run_selected_transform_cases<16>()) {
        return 4;
    }
    if (!run_selected_transform_pointer_case<4, sizeof(std::int32_t)>()) {
        return 5;
    }
    return 0;
}

File: examples/rust/src/bin/selected_transform_operator.rs
use tsl::tsl_core::StaticSimdVector;
use tsl::profile;

struct Square;

impl<V> profile::algo::UnaryKernel<V> for Square
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::MulImpl,
    V::RegisterType: Copy,
{
    fn apply(&mut self, value: V::RegisterType) -> V::RegisterType {
        profile::mul::<V>(value, value)
    }
}

struct Add;

impl<V> profile::algo::BinaryKernel<V> for Add
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::AddImpl,
{
    fn apply(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::RegisterType {
        profile::add::<V>(left, right)
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 13) % 47) as i32 - 23;
        *right_value = ((i * 7) % 31) as i32 - 15;
    }
}

fn make_selection(count: usize) -> Vec<usize> {
    let mut indices = Vec::new();
    for row in (0..count).rev() {
        if row % 3 == 0 || row % 5 == 0 {
            indices.push(row);
        }
    }
    indices
}

fn verify_unary(left: &[i32], indices: &[usize], output: &[i32], sentinel: i32) {
    for (i, &row) in indices.iter().enumerate() {
        assert_eq!(output[i], left[row] * left[row]);
    }
    for value in output.iter().skip(indices.len()) {
        assert_eq!(*value, sentinel);
    }
}

fn verify_binary(left: &[i32], right: &[i32], indices: &[usize], output: &[i32], sentinel: i32) {
    for (i, &row) in indices.iter().enumerate() {
        assert_eq!(output[i], left[row] + right[row]);
    }
    for value in output.iter().skip(indices.len()) {
        assert_eq!(*value, sentinel);
    }
}

fn main() {
    const COUNT: usize = 1003;
    const SENTINEL: i32 = 7_654_321;

    let mut left = vec![0i32; COUNT];
    let mut right = vec![0i32; COUNT];
    fill_inputs(&mut left, &mut right);
    let indices = make_selection(COUNT);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;

            let mut output = vec![SENTINEL; COUNT];
            let mut square = Square;
            profile::algo::transform_selected_unary(policy, &mut square, &left, &indices, &mut output);
            verify_unary(&left, &indices, &output, SENTINEL);

            output.fill(SENTINEL);
            let mut add = Add;
            profile::algo::transform_selected_binary(
                policy,
                &mut add,
                &left,
                &right,
                &indices,
                &mut output,
            );
            verify_binary(&left, &right, &indices, &output, SENTINEL);

            output.fill(SENTINEL);
            let mut scaled_square = Square;
            unsafe {
                profile::algo::transform_selected_unary_scaled_raw::<4, _, _, _>(
                    policy,
                    &mut scaled_square,
                    left.as_ptr(),
                    indices.as_ptr(),
                    output.as_mut_ptr(),
                    indices.len(),
                );
            }
            verify_unary(&left, &indices, &output, SENTINEL);
        }};
    }

    run_policy!(tsl::dataparallel::native());
    run_policy!(tsl::dataparallel::fixed::<1>());
    run_policy!(tsl::dataparallel::generic::<4>());
    run_policy!(tsl::dataparallel::generic::<16>());
}
